src/libraries/System.Private.CoreLib/src/System/SearchValues/ProbabilisticMapState.cs

```diff
@@ -0,0 +1,309 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+#pragma warning disable CS8500 // Takes the address of a managed type
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Stores the state necessary to call vectorized members on <see cref="ProbabilisticMap"/>,
+    /// as well as (optionally) a precomputed perfect hash table for faster single-character lookups/match confirmations.
+    /// When the hash table isn't available, the structure stores a pointer to the span of values in the set instead.
+    /// </summary>
+    internal unsafe struct ProbabilisticMapState
+    {
+        private const int MaxModulus = char.MaxValue + 1;
+
+        public ProbabilisticMap Map;
+
+        // Hash entries store each value from the set at the index determined by the remainder modulo the table size.
+        // As every value has a unique remainder, we can check if a value is contained in the set by checking
+        // _hashEntries[value % _hashEntries.Length] == value (see FastContains below).
+        // The multiplier is used for faster modulo operations when determining the hash table index.
+        // Exactly one of _hashEntries and _slowContainsValuesPtr may be initialized at the same time.
+        private readonly uint _multiplier;
+        private readonly char[]? _hashEntries;
+        private readonly ReadOnlySpan<char>* _slowContainsValuesPtr;
+
+        public ProbabilisticMapState(ReadOnlySpan<char> values, int maxInclusive)
+        {
+            Debug.Assert(!values.IsEmpty);
+
+            Map = new ProbabilisticMap(values);
+
+            uint modulus = FindModulus(values, maxInclusive);
+            _multiplier = GetFastModMultiplier(modulus);
+            _hashEntries = new char[modulus];
+
+            // Some hash entries will remain unused.
+            // We can't leave them uninitialized as that would lead to false positives for values divisible by modulus.
+            _hashEntries.AsSpan().Fill(values[0]);
```

Comment:
```suggestion
            // Some hash entries will remain unused.
            // We must ensure the first entry is initialized, as it would otherwise erroneously match (char)0.
            // The exact value doesn't matter, as long as it's in the set of our values.
            _hashEntries[0] = values[0];
```

While still functionally correct, the comment was a bit misleading. This would only be a problem for `(char)0`, not other values divisible by modulus (anything other than 0 wouldn't pass the `entries[...] == value` check as that entry is 0).